File: go/977/main.go
package main

import "fmt"

func main() {
	nums := []int{-4, -1, 0, 3, 10}
	fmt.Println(sortedSquares(nums))
}

func sortedSquares(nums []int) []int {
	res := make([]int, len(nums))
	i := 0
	j := len(nums) - 1
	cur := j
	for i <= j {
		if abs(nums[i]) >= abs(nums[j]) {
			res[cur] = nums[i] * nums[i]
			i++
		} else {
			res[cur] = nums[j] * nums[j]
			j--
		}
		cur--
	}
	return res
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

File: c/977/main.c
#include <stdlib.h>
#include <stdio.h>

int* sortedSquares(int* nums, int numsSize, int* returnSize) {
    int* res = (int*)malloc(sizeof(int) * numsSize);
    *returnSize = numsSize;
    int i = 0;
    int j = numsSize - 1;
    while (i <= j) {
        if (abs(nums[i]) >= abs(nums[j])) {
            res[--numsSize] = nums[i] * nums[i];
            i++;
        } else {
            res[--numsSize] = nums[j] * nums[j];
            j--;
        }
    }
    return res;
}

int main() {
    int nums[] = {-4, -1, 0, 3, 10};
    int returnSize = 0;
    int* sorted = sortedSquares(nums, sizeof(nums) / sizeof(int), &returnSize);
    for (int i = 0; i < returnSize; i++) {
        printf("%d ", sorted[i]);
    }
    free(sorted);
}
